contributors/design-proposals/encryption.md

```diff
@@ -0,0 +1,205 @@
+rotate API endpoint eventually. To account for failure scenarios during key rotation, the old and new DEK will be stored in etcd during the rotation.
+
+## Master Configuration
+In order to enable encryption, a user must first create a KEK DB file and tell the API server to use it with --key-encryption-key-db-path=/path/to/kekdb/file. The file will be a simple YAML file that lists all of the keys:
+
+```yaml
+kind: KeyEncryptionKeyDatabase
+version: v1
+keys:
+ - foo
+ - bar
+ - baz
+```
+
+The user will also need to specify the encryption provider and the resources to encrypt as follows:
+```yaml
+--key-encryption-key-db-path=/path/to/key-encryption-key/db
+--encryption-provider=default
+--encrypt-resource=v1/Secrets
+--encrypt-resource=v1/ConfigMap
+```
+
+As mentioned previously, the first pass at encryption will only support the default encryption provider.
+
+## Additional Considerations
+
+### Other encryption providers
+
+It should be easy to introduce new variations on the basic provider - such as those that require a password via STDIN on startup to decrypt the keys on disk, or wish to exchange a temporary token on disk for the actual keys from a third party security server.
+
+We expect that the provider implementation may therefore be layered or composed and that as much as possible providers should be separated behind interfaces.
+
+### Potential providers
+
+* A local HSM implementation that retrieves the keys from the secure enclave prior to reusing the AES-GCM implementation (initialization of keys only)
+* Exchanging a local temporary token for the actual decryption tokens from a networked secret vault
+* Decrypting the AES-256 keys from disk using asymmetric encryption combined with a user input password
+* Sending the data over the network to a key management system for encryption and decryption (Google KMS, Amazon KMS, Hashicorp Vault w/ Transit backend)
+
+### Backwards Compatibility
+Once a user encrypts any resource in etcd, they are locked to that Kubernetes version and higher unless they choose to manually decrypt that resource in etcd. This will be discouraged. It will be highly recommended that users discern if their Kubernetes cluster is on a stable version before enabling encryption.
+
+### Performance
+
+Introducing even a relatively well tuned AES-GCM implementation is likely to have performance implications for Kubernetes.
+Fortunately, existing optimizations occur above the storage layer and so the highest penalty will be incurred on writes when secrets are created or updated.
+In multi-tenant Kube clusters secrets tend to have the highest load factor (there are 20-40 resources types per namespace, but most resources
+only have 1 instance where secrets might have 3-9 instances across 10k namespaces).  Writes are uncommon, creates usually happen
+only when a namespace is created, and reads are somewhat common.
+
```

Comment:
In the meeting @smarterclayton indicated that there are two secrets access patterns that are most common. I think it was getting and listing. Can we include that so that alternate encryption provider implementers can see why keeping most encryption/decryption operations in process in the API server is preferable?